lib/web/apiserver.go

```diff
@@ -3726,12 +3747,103 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 	})
 }
 
+// WithClusterAuthWS wraps a ClusterWebsocketHandler to ensure that a request is authenticated
+// to this proxy via websocket if websocketAuth is true, or via query parameter if false (the same as WithAuth), as
+// well as to grab the remoteSite (which can represent this local cluster or a remote trusted cluster)
+// as specified by the ":site" url parameter.
+//
+// TODO(lxea): remove the 'websocketAuth' bool once the deprecated websocket handlers are removed
+func (h *Handler) WithClusterAuthWS(websocketAuth bool, fn ClusterWebsocketHandler) httprouter.Handle {
+	return httplib.MakeHandler(func(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+		if websocketAuth {
+			sctx, ws, site, err := h.authenticateWSRequestWithCluster(w, r, p)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			defer ws.Close()
+
+			if _, err := fn(w, r, p, sctx, site, ws); err != nil {
+				errEnvelope := Envelope{
+					Type:    defaults.WebsocketError,
+					Payload: err.Error(),
+				}
+				env, err := errEnvelope.Marshal()
+				if err != nil {
+					h.log.WithError(err).Error("error marshaling proto")
+					return nil, nil
+				}
+				if err := ws.WriteMessage(websocket.BinaryMessage, env); err != nil {
+					h.log.WithError(err).Error("error writing proto")
+					return nil, nil
+				}
+			}
+			return nil, nil
+		}
+
+		sctx, site, err := h.authenticateRequestWithCluster(w, r, p)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		upgrader := websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+			CheckOrigin:     func(r *http.Request) bool { return true },
```

Comment:
What would be preferred here, I dont fully understand cors, would it be set like this so connections can be made to other proxies on different clusters?

Reply:
https://github.com/gravitational/teleport/pull/37520/commits/39c56a4197b26402f42081c513868752fa076681

I hunted this down to an old comment [here](https://github.com/gravitational/teleport/blob/84b64fe487f9a74c64f702b65e2dc7bca3eb42a5/lib/web/terminal.go#L192-L198), since we're maintaining the old behavior then we'll consider it out of scope for this PR. I consolidated the question into a single global upgrader in https://github.com/gravitational/teleport/pull/37520/commits/39c56a4197b26402f42081c513868752fa076681.